text/0000-rebalancing-coherence.md

```diff
@@ -0,0 +1,293 @@
+            None
+        }
+    }
+}
+```
+
+The problem here is that `Option<&LinkedListEntry>` is no longer
+considered a local type. A similar restriction would be that one
+cannot define an impl over `Box<LinkedListEntry>`; but this was not
+observed in practice.
+
+Both of these restrictions can be overcome by using a new type.  For
+example, the code above could be changed so that instead of writing
+the impl for `Option<&LinkedListEntry>`, we define a type `LinkedList`
+that wraps the option and implement on that:
+
+```rust
+struct LinkedListEntry<'a> {
+    data: i32,
+    next: LinkedList<'a>
+}
+
+struct LinkedList<'a> {
+    data: Option<&'a LinkedListEntry>
+}
+
+impl<'a> Iterator for LinkedList<'a> {
+    type Item = i32;
+    
+    fn next(&mut self) -> Option<i32> {
+        if let Some(ptr) = self.data {
+            *self = Some(ptr.next);
+            Some(ptr.data)
+        } else {
+            None
+        }
+    }
+}
+```
+
+#### Errors from cargo and the fundamental attribute
+
+We also applied our prototype to all the "Most Downloaded" cargo
+crates as well as the `iron` crate. That exercise uncovered a few
+patterns that the simple rules presented thus far can't handle.
+
+The first is that it is common to implement traits over boxed trait
+objects. For example, the `error` crate defines an impl:
+
+- `impl<E: Error> FromError<E> for Box<Error>`
+
+Here, `Error` is a local trait defined in `error`, but `FromError` is
+the trait from `libstd`. This impl would be illegal because
+`Box<Error>` is not considered local as `Box` is not local.
+
+The second is that it is common to use `FnMut` in blanket impls,
+similar to how the `Pattern` trait in `libstd` works. The `regex` crate
+in particular has the following impls:
+
+- `impl<'t> Replacer for &'t str`
+- `impl<F> Replacer for F where F: FnMut(&Captures) -> String`
+- these are in conflict because this requires that `&str: !FnMut`, and
+  neither `&str` nor `FnMut` are local to `regex`
+
+Given that overloading over closures is likely to be a common request,
+and that the `Fn` traits are well-known, core traits tied to the call
+operator, it seems reasonable to say that implementing a `Fn` trait is
+itself a breaking change. (This is not to suggest that there is
+something *fundamental* about the `Fn` traits that distinguish them
+from all other traits; just that if the goal is to have rules that
+users can easily remember, saying that implememting a core operator
+trait is a breaking change may be a reasonable rule, and it enables
+useful patterns to boot -- patterns that are baked into the libstd
+APIs.)
+
+To accommodate these cases (and future cases we will no doubt
+encounter), this RFC proposes an unstable attribute
+`#[fundamental]`. `#[fundamental]` can be applied to types and traits
+with the following meaning:
+
+- A `#[fundamental]` type `Foo` is one where implementing a blanket
+  impl over `Foo` is a breaking change. As described, `&` and `&mut` are
```

Comment:
What exactly is a "blanket impl"? Does something like `impl<T> SomeTrait for Foo<T, Bar> {...}` count?